File: javascript/test/corpus.js
/* The corpus runner.
 *
 * Reads spec/plugin.json — the COMMITTED artifact, not the aontu source
 * — exactly as every other port's runner does. No port needs a Node
 * toolchain to run its tests, and this one does not get a private door
 * into the source either.
 *
 * A group name selects the subject. That is the whole dispatch, and it
 * is deliberately dumb: a runner that inferred the subject from the
 * entry's shape would silently run the wrong function when an entry was
 * mistyped. */

'use strict'

const Fs = require('node:fs')
const Path = require('node:path')

const { codeof } = require('../src/index')

const SPEC = Path.join(__dirname, '..', '..', 'spec', 'plugin.json')

function corpus() {
  return JSON.parse(Fs.readFileSync(SPEC, 'utf8'))
}

function section(name) {
  const spec = corpus()
  const sec = spec.primary && spec.primary[name]
  if (null == sec) throw new Error('no such corpus section: ' + name)
  const out = {}
  for (const g of Object.keys(sec)) {
    if ('DEF' === g) continue
    if (sec[g] && Array.isArray(sec[g].set)) out[g] = sec[g].set
  }
  return out
}

/** A stable label, so a failure names the entry rather than an index. */
function label(group, i, e) {
  return e.id ? e.id : group + '#' + i
}

/** Deep equality over spec values. Key order never matters; list order
 * always does. */
function equal(a, b) {
  if (a === b) return true
  if (Array.isArray(a) || Array.isArray(b)) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) return false
    for (let i = 0; i < a.length; i++) if (!equal(a[i], b[i])) return false
    return true
  }
  if (isMap(a) && isMap(b)) {
    const ka = Object.keys(a).sort()
    const kb = Object.keys(b).sort()
    if (ka.length !== kb.length) return false
    for (let i = 0; i < ka.length; i++) if (ka[i] !== kb[i]) return false
    for (const k of ka) if (!equal(a[k], b[k])) return false
    return true
  }
  return false
}

/** Partial match: every key the expectation names must agree, and keys
 * it does not name are ignored. `__EXISTS__` asserts presence without
 * pinning a value; `/re/` matches a string as a regular expression. */
function matches(expect, actual) {
  if ('__EXISTS__' === expect) return undefined !== actual
  if ('__UNDEF__' === expect) return undefined === actual
  if ('__NULL__' === expect) return null === actual

  if ('string' === typeof expect && 2 < expect.length &&
      expect.startsWith('/') && expect.endsWith('/')) {
    if ('string' !== typeof actual) return false
    return new RegExp(expect.substring(1, expect.length - 1)).test(actual)
  }

  if (Array.isArray(expect)) {
    if (!Array.isArray(actual) || expect.length !== actual.length) return false
    for (let i = 0; i < expect.length; i++) if (!matches(expect[i], actual[i])) return false
    return true
  }

  if (isMap(expect)) {
    if (!isMap(actual)) return false
    for (const k of Object.keys(expect)) if (!matches(expect[k], actual[k])) return false
    return true
  }

  return expect === actual
}

function isMap(v) {
  return null != v && 'object' === typeof v && !Array.isArray(v)
}

/** Run one entry against a subject and report the disagreement, if any.
 *
 * The three combinations the spec format allows are enforced here as
 * well as at build time, because a runner that quietly accepted `err`
 * beside `out` would let a contradictory entry pass. */
function check(e, subject) {
  if (undefined !== e.err && undefined !== e.out) {
    return 'entry has both err and out'
  }

  let value
  let raised = null
  try {
    value = subject(e)
  }
  catch (err) {
    raised = err
  }

  if (undefined !== e.err) {
    if (null == raised) return 'expected a raise, got: ' + JSON.stringify(value)
    if (true !== e.err) {
      // Errors compare by CODE (§12). Message wording is a port's own
      // business, and pinning it would make every translation a corpus
      // change.
      if (codeof(raised) !== e.err) {
        return 'expected code ' + e.err + ', got ' + codeof(raised) +
          ' (' + raised.message + ')'
      }
    }
    if (undefined !== e.match) {
      const got = { err: { code: codeof(raised), message: raised.message, name: raised.name } }
      if (!matches(e.match, got)) {
        return 'error did not match ' + JSON.stringify(e.match) + ', got ' + JSON.stringify(got)
      }
    }
    return null
  }

  if (null != raised) {
    return 'unexpected raise: ' + codeof(raised) + ' ' + raised.message
  }

  if (undefined !== e.out) {
    if (!equal(e.out, value)) {
      return 'expected ' + JSON.stringify(e.out) + ', got ' + JSON.stringify(value)
    }
  }

  if (undefined !== e.match) {
    if (!matches(e.match, { in: e.in, out: value })) {
      return 'did not match ' + JSON.stringify(e.match) + ', got out=' + JSON.stringify(value)
    }
  }

  if (undefined === e.out && undefined === e.match) {
    return 'entry asserts nothing'
  }

  return null
}

module.exports = { corpus, section, label, equal, matches, check }

File: elixir/test/corpus.exs
# The corpus runner.
#
# Reads spec/plugin.json - the COMMITTED artifact, not the aontu source -
# exactly as every other port's runner does. No port needs a Node
# toolchain to run its tests, and this one does not get a private door
# into the source either.
#
# A group name selects the subject. That is the whole dispatch, and it is
# deliberately dumb: a runner that inferred the subject from the entry's
# shape would silently run the wrong function when an entry was mistyped.

defmodule Corpus do
  alias Voxgig.Plugin.Json
  alias Voxgig.Plugin.Types

  @spec_path Path.join([__DIR__, "..", "..", "spec", "plugin.json"])

  # A sentinel for "this key was not present". `Map.get` returns nil for
  # both an absent key and a JSON null, and `__UNDEF__` and `__NULL__` are
  # different assertions.
  @missing :__corpus_missing__

  def corpus, do: @spec_path |> File.read!() |> Json.parse()

  def section(spec, name) do
    sec = Types.get(Types.get(spec, "primary") || %{}, name)
    if is_nil(sec), do: raise(ArgumentError, "no such corpus section: #{name}")

    sec
    |> Enum.filter(fn {group, body} ->
      "DEF" != group and is_map(body) and is_list(Types.get(body, "set"))
    end)
    |> Map.new(fn {group, body} -> {group, Types.get(body, "set")} end)
  end

  # A stable label, so a failure names the entry rather than an index.
  def label(group, i, entry), do: Types.get(entry, "id") || "#{group}##{i}"

  @doc """
  Deep equality over spec values. Key order never matters; list order
  always does.

  Elixir's `true == 1` is already false - atoms and numbers never compare
  equal - so the bool guard the other dynamic ports need is not required
  here. `1 == 1.0` IS true, as it is in ruby and javascript, and no corpus
  entry turns on the difference.
  """
  def same(a, b) when is_map(a) and is_map(b) do
    map_size(a) == map_size(b) and
      Enum.all?(a, fn {k, v} -> Map.has_key?(b, k) and same(v, Map.get(b, k)) end)
  end

  def same(a, b) when is_list(a) and is_list(b) do
    length(a) == length(b) and a |> Enum.zip(b) |> Enum.all?(fn {x, y} -> same(x, y) end)
  end

  def same(a, b) when is_list(a) or is_list(b) or is_map(a) or is_map(b), do: false

  def same(a, b), do: a == b

  @doc """
  Partial match: every key the expectation names must agree, and keys it
  does not name are ignored. `__EXISTS__` asserts presence without pinning
  a value; `/re/` matches a string as a regular expression.
  """
  def matches("__EXISTS__", actual), do: @missing != actual and not is_nil(actual)
  def matches("__UNDEF__", actual), do: @missing == actual
  def matches("__NULL__", actual), do: @missing != actual and is_nil(actual)

  def matches(expect, @missing), do: matches(expect, nil)

  def matches(expect, actual) when is_binary(expect) do
    if 2 < byte_size(expect) and String.starts_with?(expect, "/") and
         String.ends_with?(expect, "/") do
      is_binary(actual) and
        Regex.match?(Regex.compile!(binary_part(expect, 1, byte_size(expect) - 2)), actual)
    else
      expect == actual
    end
  end

  def matches(expect, actual) when is_list(expect) do
    is_list(actual) and length(expect) == length(actual) and
      expect |> Enum.zip(actual) |> Enum.all?(fn {e, a} -> matches(e, a) end)
  end

  def matches(expect, actual) when is_map(expect) do
    is_map(actual) and
      Enum.all?(expect, fn {k, v} ->
        matches(v, if(Map.has_key?(actual, k), do: Map.get(actual, k), else: @missing))
      end)
  end

  def matches(expect, actual), do: expect == actual

  @doc """
  Run one entry against a subject and report the disagreement, if any.

  The three combinations the spec format allows are enforced here as well
  as at build time, because a runner that quietly accepted `err` beside
  `out` would let a contradictory entry pass.
  """
  def check(entry, subject) do
    if Types.has(entry, "err") and Types.has(entry, "out") do
      "entry has both err and out"
    else
      verdict(entry, outcome(entry, subject))
    end
  end

  defp outcome(entry, subject) do
    {:ok, subject.(entry)}
  rescue
    e -> {:raised, e}
  end

  defp verdict(entry, {:ok, value}) do
    cond do
      Types.has(entry, "err") ->
        "expected a raise, got: #{Types.encode(value)}"

      Types.has(entry, "out") and not same(Types.get(entry, "out"), value) ->
        "expected #{Types.encode(Types.get(entry, "out"))}, got #{Types.encode(value)}"

      Types.has(entry, "match") and
          not matches(Types.get(entry, "match"), %{"in" => Types.get(entry, "in"), "out" => value}) ->
        "did not match #{Types.encode(Types.get(entry, "match"))}, got out=#{Types.encode(value)}"

      not Types.has(entry, "out") and not Types.has(entry, "match") ->
        "entry asserts nothing"

      true ->
        nil
    end
  end

  defp verdict(entry, {:raised, e}) do
    if Types.has(entry, "err"), do: errverdict(entry, e), else:
      "unexpected raise: #{Types.codeof(e)} #{Types.message(e)}"
  end

  defp errverdict(entry, e) do
    want = Types.get(entry, "err")
    # Errors compare by CODE (section 12). Message wording is a port's own
    # business, and pinning it would make every translation a corpus
    # change.
    got = Types.codeof(e)

    cond do
      true != want and got != want ->
        "expected code #{want}, got #{got} (#{Types.message(e)})"

      Types.has(entry, "match") and
          not matches(Types.get(entry, "match"),
                      %{"err" => %{"code" => got, "message" => Types.message(e),
                                   "name" => "PluginError"}}) ->
        "error did not match #{Types.encode(Types.get(entry, "match"))}, got code=#{got}"

      true ->
        nil
    end
  end
end
